staticd/static_vty.c

```diff
@@ -367,169 +158,259 @@ static int static_route_leak(
+		/* distance,tag,table-id processing */
+
+		if (src_str)
+			snprintf(xpath_list, sizeof(xpath_list),
+				 FRR_S_ROUTE_SRC_INFO_KEY_XPATH,
+				 "frr-staticd:staticd", "staticd", svrf,
+				 buf_prefix, buf_src_prefix, distance, tag,
+				 table_id);
+		else
+			snprintf(xpath_list, sizeof(xpath_list),
+				 FRR_STATIC_ROUTE_INFO_KEY_XPATH,
+				 "frr-staticd:staticd", "staticd", svrf,
+				 buf_prefix, distance, tag, table_id);
+
+		nb_cli_enqueue_change(vty, xpath_list, NB_OP_CREATE, NULL);
+
+		/* nexthop processing */
+
+		if (src_str) {
+			snprintf(xpath_list, sizeof(xpath_list),
+				 FRR_S_ROUTE_SRC_NH_KEY_XPATH,
+				 "frr-staticd:staticd", "staticd", svrf,
+				 buf_prefix, buf_src_prefix, distance, tag,
+				 table_id, buf_nh_type, nh_svrf, buf_gate_str,
+				 ifname);
+		} else {
+			snprintf(xpath_list, sizeof(xpath_list),
+				 FRR_STATIC_ROUTE_NH_KEY_XPATH,
+				 "frr-staticd:staticd", "staticd", svrf,
+				 buf_prefix, distance, tag, table_id,
+				 buf_nh_type, nh_svrf, buf_gate_str, ifname);
+		}
+		nb_cli_enqueue_change(vty, xpath_list, NB_OP_CREATE, NULL);
+
+		if (type == STATIC_BLACKHOLE) {
+			if (src_str)
+				snprintf(xpath_list, sizeof(xpath_list),
+					 FRR_S_ROUTE_SRC_NH_BH_XPATH,
+					 "frr-staticd:staticd", "staticd", svrf,
+					 buf_prefix, buf_src_prefix, distance,
+					 tag, table_id, buf_nh_type, nh_svrf,
+					 buf_gate_str, ifname);
 			else
-				zlog_warn(
-					"%s: %s: Nexthop interface name can not be from reserved keywords (Null0, reject, blackhole)",
-					__func__, dest_str);
-			return CMD_WARNING_CONFIG_FAILED;
+				snprintf(xpath_list, sizeof(xpath_list),
+					 FRR_STATIC_ROUTE_NH_BH_XPATH,
+					 "frr-staticd:staticd", "staticd", svrf,
+					 buf_prefix, distance, tag, table_id,
+					 buf_nh_type, nh_svrf, buf_gate_str,
+					 ifname);
+
+			/* Route flags */
+			if (flag_str) {
+				switch (flag_str[0]) {
+				case 'r':
+					bh_type = "reject";
+					break;
+				case 'b':
+					bh_type = "unspec";
+					break;
+				case 'N':
+					bh_type = "null";
+					break;
+				default:
+					bh_type = NULL;
+					break;
+				}
+				nb_cli_enqueue_change(vty, xpath_list,
+						      NB_OP_MODIFY, bh_type);
+			} else {
+				nb_cli_enqueue_change(vty, xpath_list,
+						      NB_OP_DESTROY, NULL);
+			}
 		}
-	}
 
-	/* Route flags */
-	if (flag_str) {
-		switch (flag_str[0]) {
-		case 'r':
-			bh_type = STATIC_BLACKHOLE_REJECT;
-			break;
-		case 'b':
-			bh_type = STATIC_BLACKHOLE_DROP;
-			break;
-		case 'N':
-			bh_type = STATIC_BLACKHOLE_NULL;
-			break;
-		default:
-			if (vty)
-				vty_out(vty, "%% Malformed flag %s \n",
-					flag_str);
+		if (type == STATIC_IPV4_GATEWAY_IFNAME
+		    || type == STATIC_IPV6_GATEWAY_IFNAME) {
+
+			if (src_str)
+				snprintf(xpath_list, sizeof(xpath_list),
+					 FRR_S_ROUTE_SRC_NH_ONLINK_XPATH,
+					 "frr-staticd:staticd", "staticd", svrf,
+					 buf_prefix, buf_src_prefix, distance,
+					 tag, table_id, buf_nh_type, nh_svrf,
+					 buf_gate_str, ifname);
```

Comment:
Actually, i can't use this relative framework for staticd. 
```                if (vty->xpath_index > 0
                    && ((xpath_base_fmt && xpath_base[0] == '.') 
                        || change->xpath[0] == '.'))
                        strlcpy(xpath, VTY_CURR_XPATH, sizeof(xpath));
This code modifies the xpath in case of vrf.
c t
vrf red      ------------------ index 0
ip route 1.1.1.1/32 ens192 --- index 1

Here i don't want this relative mapping, as staticd is not augmented from frr-vrf.yang model.
But I have removed few macros and make the code generic.

Reply:
What I'm trying to say is that you're doing this for a command like "ip route 60.0.0.0/8 10.0.1.3":
```
nb_cli_enqueue_change():
  xpath: "/frr-routing:routing/control-plane-protocols/control-plane-protocol[type='frr-staticd:staticd'][name='staticd'][vrf='default']/frr-staticd:staticd/route-list[prefix='60.0.0.0/8'][afi-safi='frr-routing:ipv4-unicast']/path-list[distance='1'][tag='0'][table-id='0']"

nb_cli_enqueue_change():
  xpath: "/frr-routing:routing/control-plane-protocols/control-plane-protocol[type='frr-staticd:staticd'][name='staticd'][vrf='default']/frr-staticd:staticd/route-list[prefix='60.0.0.0/8'][afi-safi='frr-routing:ipv4-unicast']/path-list[distance='1'][tag='0'][table-id='0']/frr-nexthops/nexthop[nh-type='ip4'][vrf='default'][gateway='10.0.1.3'][interface='(null)']"

nb_cli_apply_changes():
  xpath_base: "/frr-routing:routing/control-plane-protocols/control-plane-protocol[type='frr-staticd:staticd'][name='staticd'][vrf='default']/frr-staticd:staticd/route-list[prefix='60.0.0.0/8'][afi-safi='frr-routing:ipv4-unicast']/path-list[distance='1'][tag='0'][table-id='0']"
```

Whereas you could do this:
```
nb_cli_enqueue_change():
  xpath: "."

nb_cli_enqueue_change():
  xpath: "./frr-nexthops/nexthop[nh-type='ip4'][vrf='default'][gateway='10.0.1.3'][interface='(null)']"

nb_cli_apply_changes():
  xpath_base: "/frr-routing:routing/control-plane-protocols/control-plane-protocol[type='frr-staticd:staticd'][name='staticd'][vrf='default']/frr-staticd:staticd/route-list[prefix='60.0.0.0/8'][afi-safi='frr-routing:ipv4-unicast']/path-list[distance='1'][tag='0'][table-id='0']
```
(the first change (`"."`) is actually redundant since adding a nexthop will create that xpath anyway)

The base xpath you're passing to `nb_cli_apply_changes()` is being ignored since none of your enqueued changes use relative paths. You can change that to simplify the code considerably (i.e. no more repeating those lengthy snprints multiple times).

Reply:
I noticed several changes were done to this function but it still isn't using relative paths as I suggested. In any case, it looks a bit better now. We can change it again later in the future if necessary, so this is resolved from my perspective.